### ISSUE TITLE
[TypeScript] styled argument do not accept string for tagnames

### ISSUE THREAD

**Issue by renatorib** [styled-components/styled-components#1080]
It does not work:

![image](https://user-images.githubusercontent.com/3277185/29194229-0ec54c6a-7dff-11e7-8e96-1a07ecf01954.png)

It works:

![image](https://user-images.githubusercontent.com/3277185/29194326-6752f684-7dff-11e7-93e5-e6577ee8ecb4.png)

Following `styled typings` I found that first argument receive `Component<P>`:
```tsx
type Component<P> = ComponentClass<P> | StatelessComponent<P>;
```

If I add `| string;` at Component type, the first example works successfully.

What I can do in this case? And why Component argument do not accept string?

**Comment by philpl** [comment]
This does look like an oversight to me. Rest assured that you can still use the shorthands. Please check out our docs for some examples of them: https://www.styled-components.com/docs

cc @styled-components/typers

**Comment by renatorib** [comment]
Sorry, I guess I did not express the problem very well.

I want to create styled components with prop types and using string for html tags

```tsx
const Comp = styled<Props>('div')``;
//                 ^ [ts] Argument of type "'div'" is not assignable to parameter of type 'Component<Props>'
```

But typescript does not let me do that, because of this type: 
```tsx 
type Component<P> = ComponentClass<P> | StatelessComponent<P>;
```

**Comment by renatorib** [comment]
Ops, I've closed issue by missclick. Sorry.

**Comment by Igorbek** [comment]
I'm sure #1067 should be addressing this feature.

**Comment by renatorib** [comment]
@Igorbek exactly that issue.

For now, I've created a wrapper to handle this:
``` tsx
import React, { StatelessComponent, ComponentClass } from 'react';
import styled, { Interpolation, ThemedStyledProps } from 'styled-components';

function styledTyped<P>(
  Component: StatelessComponent | ComponentClass | string,
) {
  return (
    strings: TemplateStringsArray,
    ...interpolations: Interpolation<ThemedStyledProps<P, any>>[],
  ) => {
    return styled<P>(() => <Component />)(strings, ...interpolations);
  };
}

// after

type LineProps = {
  background: string;
};

const Line = styledTyped<LineProps>('div')`
  width: 100%;
  height: 1px;
  background: ${p => p.background};
`;
```

**Comment by philpl** [comment]
@renatorib is there any reason why you're using this workaround instead of the shorthands?

**Comment by renatorib** [comment]
@philpl because I can't pass types using shorthands.

You know how?

**Comment by renatorib** [comment]
This way, styled do not recognize props in ThemedStyledProps (because I don't pass any prop types definitions):
![image](https://user-images.githubusercontent.com/3277185/29228263-864bd794-7eaf-11e7-8658-486beb161e14.png)

Using the workaround it recognize, but I can't use shorthands as well:

![image](https://user-images.githubusercontent.com/3277185/29228343-c9617480-7eaf-11e7-8a87-4760aa672c9b.png)

![image](https://user-images.githubusercontent.com/3277185/29228356-cf717be0-7eaf-11e7-89b3-91aadb700d2b.png)

**Comment by philpl** [comment]
@renatorib before—I don't know if it still works—TS was able to infer it back to the generic from usage:

```ts
interface Props {
  color: string;
}

styled.div`
  color: ${(p: Props) => p.color};
`
```

But if that doesn't work you might still be able to do this:

```ts
styled.div`
  color: ${(p: Props) => p.color};
` as StyledFunction<Props>;
```

**Comment by renatorib** [comment]
Yes, first example work. But I'll need to pass `(p: Props)` whenever I use props in my styled component. Seems overkill to me. Imagine that I've a large/complex styled component, and will need to repeat it several times. The best way is when interpolations function from template strings already receive the props in the callback argument.

(but it is also a solution as well)

I also think this workaround is better because it seems will be same 'api' to the future, according to work in #1067: 

![image](https://user-images.githubusercontent.com/3277185/29229825-bb76ba90-7eb6-11e7-8113-2bb6ec8c812e.png)

**Comment by Igorbek** [comment]
it allows to pass that props just once:

```ts
const Title = styled.h1`
  color: ${(props: Props) => props.color};
  font-size: ${props => props.size /* <- here it is inferred */ }
`;
```

**Comment by renatorib** [comment]
@Igorbek hummmmm 🤔  now it's interesting

**Comment by philpl** [comment]
@renatorib please do not just close the issue, since we actually identified a bug here ;)

**Comment by renatorib** [comment]
Oh, okay.

**Comment by renatorib** [comment]
Closing since it's fixed in #1139
